packages/back-end/src/queryRunners/QueryRunner.ts

```diff
@@ -656,29 +652,31 @@ export abstract class QueryRunner<
     logger.debug("Creating query for: " + name);
     const concurrencyLimitReached = await this.concurrencyLimitReached();
     const dependenciesComplete = dependencies.length === 0;
-    const readyToRun =
-      dependenciesComplete && !runAtEnd && !concurrencyLimitReached;
+    const readyToRun = dependenciesComplete && !runAtEnd;
     const doc = await createNewQuery({
       query,
       queryType,
+      dependencies,
+      runAtEnd,
       datasource: this.integration.datasource.id,
       organization: this.integration.context.org.id,
       language: this.integration.getSourceProperties().queryLanguage,
-      dependencies: dependencies,
-      running: readyToRun,
-      runAtEnd: runAtEnd,
+      running: readyToRun && !concurrencyLimitReached,
     });
 
+    // Update heartbeat for the query once every 30 seconds
+    // This lets us detect orphaned queries where the thread died
+    this.updateHeartbeatOnInterval(doc);
+
     logger.debug("Created new query " + doc.id + " for " + name);
-    if (readyToRun) {
+    if (readyToRun && !concurrencyLimitReached) {
       this.executeQuery(doc, run, process);
-    } else if (dependenciesComplete && !runAtEnd) {
-      this.queuedQueryTimers[doc.id] = setTimeout(() => {
-        this.executeQueryWhenReady(doc, run, process);
-      }, INITIAL_CONCURRENCY_TIMEOUT);
```

Comment:
Was this `else if` block removed on purpose?

Reply:
Yes, this got folded into the existing else block with a slight shuffle of the condition to make it easier to follow the logic.

Reply:
I'm not 100% following the code here, so hard for me to see where the `dependenciesComplete && !runAtEnd` check has moved to.

The only time we use `runAtEnd` is for DataBricks when pipeline mode is enabled.  I believe the intention is that it's a cleanup query that we want to run at the end no matter what, even if the other queries failed.  It's pretty hard to get into this state to test, so just flagging this as something we should try and verify to make sure there's not a regression.

Reply:
Here's the relevant part of this code block for this if/else logic:

```
const readyToRun = dependenciesComplete && !runAtEnd;
```
defines `readyToRun` as to whether this query wants to be executed as soon as concurrency is available. Then once the query doc is instantiated we hit this if/else (which used to be if/elif/else) for

`if (readyToRun && !concurrencyLimitReached)`, i.e. if we're 100% good to run it now, and then the else.
Within the else, the new `if` replaces the `else if` that this comment thread is tagged on:
```
if (readyToRun && concurrencyLimitReached) {
```
which is equivalent to `dependenciesComplete && !runAtEnd && concurrencyLimitReached`, i.e. we want to start it but there's a concurrency limit.

The reason I folded it into the `else` rather than having an `else if` for it is that the previous line needs to be set for these queries as well:
```
this.runCallbacks[doc.id] = { run, process };
```

Reply:
Ah, ok.  I think I follow now.